docs/intro.rst

```diff
@@ -25,9 +25,9 @@ Let see how the library can handle this situation:
     from url_matcher import URLMatcher, Patterns
 
     matcher = URLMatcher()
-    matcher.add_or_update("us_proxy", Patterns(["site1.com"]))
-    matcher.add_or_update("uk_proxy", Patterns(["site2.com/uk"]))
-    matcher.add_or_update("ie_proxy", Patterns(["site2.com/ie"]))
+    matcher.add_or_update("us_proxy", Patterns(("site1.com",)))
```

Comment:
I liked the previous API more :( Here it is easier to make a mistake than before. 

Also, see https://stackoverflow.com/a/1708610/114795 - I'm in this camp :) This affects signature of Pattern.get_* methods as well - for them it makes a lot of sense to return lists, not tuples.

Do you think it'd be possible to have `Pattern.__init__` arguments as lists, and function return values as lists, but keep attribute values as tuples? Sorry for a lot of nitpicking :)

Reply:
I still prefer having immutable patterns if we can push throught it, but an alternative to this could be to create a function which deduplicates Pattern objects; it'd be less pure, but probably won't involve so many tiny caveats.

Reply:
> I liked the previous API more :( Here it is easier to make a mistake than before.

Agreed, most especially on one-element tuples as it's quite hard to properly remember a trailing `,` char at the end.

> Do you think it'd be possible to have `Pattern.__init__` arguments as lists, and method return values as lists, but keep attribute values as tuples, so that Pattern is still immutable?

After exploring some various ways on how to approach this tricky situation, I believe this would be the cleanest one we could have:

```python
from dataclasses import dataclass, replace
from typing import Tuple, List

@dataclass(init=False, frozen=True)
class Foo:
    data: Tuple[str]

    def __init__(self, data: List[str]):
        object.__setattr__(self, "data", tuple(data))

f = Foo(["hi", "Hello"])

f.data.append("hey") # not allowed since tuples are immutable
f.data = ("new",)    # FrozenInstanceError: cannot assign to field 'data'
f.new_field = False  # FrozenInstanceError: cannot assign to field 'new_field'

new_f = replace(f, data=["new"])  # only way would be to create a new instance
```

Implemented this change on https://github.com/zytedata/url-matcher/pull/4/commits/9d9df96c38aa85bb0aba85af6660e0d42dc102bd.

This is in line with how Python's own **dataclasses** library work: https://github.com/python/cpython/blob/main/Lib/dataclasses.py#L1119-L1122.

Let me know what you think. :) 

> but an alternative to this could be to create a function which deduplicates Pattern objects; it'd be less pure, but probably won't involve so many tiny caveats

That's true, but using such function won't be as Pythonic as simply running `set()` over a collection of `Patterns` instances. Users have to specifically remember such function or look up the API yet again